compiler/src/iree/compiler/InputConversion/MHLO/ConvertMHLOToLinalgExt.cpp

```diff
@@ -470,6 +471,125 @@ struct ReverseOpConversion : public OpConversionPattern<mhlo::ReverseOp> {
+    SmallVector<AffineMap> indexingMaps{mapIdentity};
+    SmallVector<StringRef> iterators{inputRank, "parallel"};
+    Value inputIndices =
+        rewriter
+            .create<linalg::GenericOp>(
+                loc, /*resultTensorTypes=*/initAcc.getType(),
+                /*inputs=*/ValueRange{}, /*outputs=*/ValueRange{initAcc},
+                indexingMaps, iterators,
+                [=](OpBuilder &b, Location loc, ValueRange args) {
+                  // Indicies increase along the k dimension
+                  Value index = b.create<linalg::IndexOp>(loc, kDim);
+                  Value intIndex =
+                      b.create<arith::IndexCastOp>(loc, b.getI32Type(), index);
+                  b.create<linalg::YieldOp>(loc, intIndex);
+                })
+            .getResult(0);
+
+    // Create and initialize output tensors for LinalgExt TopK results
+    // Define the output types based on the results of CHLO TopK
+    auto outputValuesType = op.getResultTypes()[0].dyn_cast<ShapedType>();
+    auto outputIndicesType = op.getResultTypes()[1].dyn_cast<ShapedType>();
+    dynSizes.clear();
+    for (auto en : llvm::enumerate(outputValuesType.getShape())) {
+      if (en.value() == ShapedType::kDynamicSize) {
+        dynSizes.push_back(
+            rewriter.create<tensor::DimOp>(loc, adaptor.operand(), en.index()));
+      }
+    }
+    Type valueElementType = outputValuesType.getElementType();
+    Type indicesElementType = outputIndicesType.getElementType();
+    Value initTensorOutputValues = rewriter.create<mlir::linalg::InitTensorOp>(
+        loc, dynSizes, outputValuesType.getShape(), valueElementType);
+    Value initTensorOutputIndices = rewriter.create<mlir::linalg::InitTensorOp>(
+        loc, dynSizes, outputIndicesType.getShape(), indicesElementType);
+    // Initialize indices to 0 and values to negative infinity
+    Attribute negInfAttr;
+    if (valueElementType.isa<IntegerType>()) {
+      negInfAttr = rewriter.getIntegerAttr(
+          valueElementType,
+          APInt::getSignedMinValue(
+              valueElementType.cast<IntegerType>().getWidth()));
+    } else {
+      auto negApFloat = APFloat::getInf(
+          valueElementType.cast<FloatType>().getFloatSemantics(),
+          /*Negative=*/true);
+      negInfAttr = rewriter.getFloatAttr(valueElementType, negApFloat);
+    }
+    Value negInf = rewriter.create<arith::ConstantOp>(loc, negInfAttr);
+    Value zero = rewriter.create<arith::ConstantOp>(
+        loc, rewriter.getZeroAttr(indicesElementType));
+    Value negInfTensor =
+        rewriter.create<linalg::FillOp>(loc, negInf, initTensorOutputValues)
+            .getResult(0);
+    Value zeroTensor =
+        rewriter.create<linalg::FillOp>(loc, zero, initTensorOutputIndices)
+            .getResult(0);
+
+    // Replace the CHLO TopK with LinalgExt TopK
+    auto topkOp = rewriter.replaceOpWithNewOp<IREE::LinalgExt::TopkOp>(
```

Comment:
Sorry for digging this up again, but can you remind me why we need a separate `indices` input tensor. That seems like information already carried by the op and does not need to be explicitly instantiated. That increases the memory foot print by a factor of 2. I know thats how the `linalg_ext.topk` is defined, but I am kind of revisiting that cause I didnt connect it till I see the lowering here.

Reply:
To clarify the arguments here: the tensor `linalg_ext::topk` needs 2 inputs and 2 outputs (though technically, they're all provided as inputs making 4 total): 
- input values and input indices
- output values and output indices

The existing `chlo::topk` op only has only a single relevant argument, the input values. See: https://github.com/tensorflow/mlir-hlo/blob/master/include/mlir-hlo/Dialect/mhlo/IR/chlo_ops.td#L768

So, for my version, I need to create the other 3 tensors and provide them as input:
- input indices- are initialized and filled using the linalg generic and index operations
- output values and indices are initialized after and filled with neg-inf and zero respectively

To answer @MaheshRavishankar  specifically, the reason I keep a separate indices tensor is to support splitting topk. I need the original indices so that I can preserve the order when merging results. I'm aware this increases memory usage. Its possible we may be able to relax this requirement for the transformer models we're working on, but at the moment we don't have concrete info that its possible. 

Does this answer your questions or am I missing something?

Reply:
Even for splitting top-K you dont need to create a separate indices tensor. I assume here you mean splitting the reduction dimension. This comes down to tiling. At every iteration (tiled or untiled) you know the current index value. 

For example [here](https://github.com/google/iree/blob/e32cc76952d37a14f73c8a7da889edf47fcc2fce/llvm-external-projects/iree-dialects/test/Dialect/iree_linalg_ext/tiling.mlir#L83), %ARG4 is the start position of the index value for that tile. You get the index for an element by %ARG4 + index within the tile. There shouldnt be a need to compute this.

Reply:
Yes, in the tiling specific case, the loop bounds provide the index for accessing the original value. 

I haven't worked out the full process for split reduction, but the initial assumption is that the individual parallel reductions and the merge step would be separate operations and wouldn't necessarily share loop bounds or the input tensor. Once these values have been shuffled around by the parallel top k, I don't think there is any way for me to derive the original indices of the values when I merge even if I'm reusing the same input tensors. Perhaps I'm missing something?

Reply:
This is where seeing the split of reductions as tiling helps. When you are tiling the reductions, at every step you know what your position in the untiled iteration space is by construction. So again you wouldnt need a separate operand for that.

Reply:
Sounds reasonable to me. The indices can be inferred in the tiling cases as well. We'll get the offsets/ivs when implementing tiling method.

Reply:
Let me expand on this a bit with just a reduction on memrefs

```
linalg.generic {
    indexing_maps = [affine_map<(d0) -> (d0)>, affine_map<(d0) -> ()>],
    iterator_types = ["reduction"]}
    ins(%source : memref<?xf32>) outs(%zero : memref<f32>) { ... }
```

When you tile this you get

```
scf.for %iv0 = %c0 to %ub step %ts {
  %1 = subview %source[%iv0][%ts][1]
   %2 = linalg.generic ... ins(%1)
}
```

Lets lower that to loops you get

```
scf.for %iv0 = %c0 to %ub step %ts {
  %1 = subview %source[%0][%ts][1]
   scf.for %iv1 = %c0 to %ts step %c1 {
       load %1[%iv1]
   }
}
```

Within the innermost loop `%iv0 + %iv1` gives you the position in the original iteration space. This value for all iterations is what is being stored in `input_indices` and is unnecessary. You have the value always, irrespective of whether you distribute the reduction dimension or not....

Reply:
The topk op works a bit different than a normal reduction since it needs to return the original indices of the top elements. This information would be lost when doing a tree implementation of topk. 
It is true that the index could be skipped for the first level of topk and to be efficient we should probably support it in which case we would either need to two ops or have the indices input be optional and if it is not there just relay on the loop induction variable.
We do need the existing version but if we were to also support a case where the induction variable is used that's what we would use for lowering from chlo. So it might be worth having a quick chat about it. My gut feeling is that in term of performance this is most likely going to be needed, in which case we may want to go back to the design of topk to support this.

Reply:
That being said, I don't think it is a reason to block this PR.

Reply:
> The topk op works a bit different than a normal reduction since it needs to return the original indices of the top elements. This information would be lost when doing a tree implementation of topk.

No it wouldnt be.

> It is true that the index could be skipped for the first level of topk and to be efficient we should probably support it in which case we would either need to two ops or have the indices input be optional and if it is not there just relay on the loop induction variable. We do need the existing version but if we were to also support a case where the induction variable is used that's what we would use for lowering from chlo.

Having two ops is definitely worth it. It just adds complexity. We definitely want to have as small as an op surface as possible. Given I am still not convinced that `input_indices` is needed, I dont see a reason for that op.

>  So it might be worth having a quick chat about it. My gut feeling is that in term of performance this is most likely going to be needed, in which case we may want to go back to the design of topk to support this.

I'd rather not have such development loops. I agree it was my mistake to not catch this on the op definition of `linalg_ext.topk`. I should have raised it at that time. Still I dont think its too late backtrack. For now the change is just a few fixes to op definition and lowering to loops/tiling (all should be minimal.

> That being said, I don't think it is a reason to block this PR.

Its a judgement call. I should have raised it in an earlier PR, but doesnt mean that we continue along this path because it was missed on initial review.

Reply:
@ThomasRaoux @MaheshRavishankar and @KoolJBlack discussed this further offline. The decision has been to make the `%input_indices` variable of TopK optional. I've created an issue for this task: https://github.com/google/iree/issues/9204

With this in mind, I think we can resolve this conversation for now. Since its not specifically blocking here,  I propose we commit this PR so that we'll have an e2e implementation of Topk. I will followup with a PR modifying the entire e2e implementation with the additional optional use case.

Reply:
My (strong) preference here is that this PR be held off until the changes to the semantics of `linalg_ext.topk` are made and all the transformations with the changes are verified. This PR would need to change to not generate the `linalg.generic` that is just materializing indices. Given that this PR is essentially generating two ops, and one of them is unnecessary, it seems better to land this after all changes to the `linalg_ext.topk`.
Since Kojo mentioned that it will be easier for him to make progress on fixing things up after this PR lands, I am not going to block it, but landing this seems premature to me.

Reply:
We're exploring how it works at this stage, so making it optional sounds good to me. We might get some perf win if the shape is static. Because we don't have to compute them during execution. 

It's always easier to me when things not getting complicated. After thinking a while, it's fine to land the PR firstly because I assume that we only need small changes when adding optional support. E.g., adding an if-else statement to decide whether we need indices operand. I'll review it later today and let Kojo make the call.

Reply:
I've gone ahead and sent out https://github.com/google/iree/pull/9228. I think this change is fairly straightforward, so if you want to review that first, I can commit that then this.

Reply:
https://github.com/google/iree/pull/9228 has merged and this change has been updated to on use a single input.